src/docfx/docset/Repository.cs

```diff
@@ -41,5 +44,26 @@ public static Repository CreateFromFolder(string path)
             }
             return new Repository(remote, branch, commit, PathUtility.NormalizeFolder(repoPath));
         }
+
+        public static (Repository repo, string pathToRepo) GetRepository(Document document)
+        {
+            var fullPath = PathUtility.NormalizeFile(System.IO.Path.Combine(document.Docset.DocsetPath, document.FilePath));
+            var repo = GetRepository(fullPath);
+            if (repo == null)
+                return default;
+
+            return (repo, PathUtility.NormalizeFile(System.IO.Path.GetRelativePath(repo.Path, fullPath)));
+        }
+
+        public static Repository GetRepository(string fullPath)
+        {
+            if (GitUtility.IsRepo(fullPath))
+                return CreateFromFolder(fullPath);
+
+            var parent = System.IO.Path.GetDirectoryName(fullPath);
+            return !string.IsNullOrEmpty(parent)
+                ? _repositoryByFolder.GetOrAdd(parent, GetRepository)
+                : null;
+        }
```

Comment:
Move these static methods to `RepositoryUtility`? No need to include global _repositoryByFolder in a specific repository instance.

Reply:
great idea